src/native/minipal/log.c

```diff
@@ -0,0 +1,329 @@
+
+// Android defines its internal log buffer used in __android_log_vprint to 1024 bytes.
+// Use same internal stack buffer size avoiding dynamic memory allocation for majority of logging.
+#define MINIPAL_LOG_BUF_SIZE 1024
+
+static int android_log_flag(minipal_log_flags flags)
+{
+    switch(flags)
+    {
+    case minipal_log_flags_fatal:
+        return ANDROID_LOG_FATAL;
+    case minipal_log_flags_error:
+        return ANDROID_LOG_ERROR;
+    case minipal_log_flags_warning:
+        return ANDROID_LOG_WARN;
+    case minipal_log_flags_info:
+        return ANDROID_LOG_INFO;
+    case minipal_log_flags_debug:
+        return ANDROID_LOG_DEBUG;
+    case minipal_log_flags_verbose:
+        return ANDROID_LOG_VERBOSE;
+    default:
+        return ANDROID_LOG_UNKNOWN;
+    }
+}
+
+static size_t log_write(minipal_log_flags flags, const char* msg, size_t msg_len)
+{
+    return __android_log_write(android_log_flag(flags), MINIPAL_LOG_RUNTIME_TAG, msg) == 1 ? msg_len : 0;
+}
+
+int minipal_log_print(minipal_log_flags flags, const char* fmt, ... )
+{
+    va_list args;
+    va_start(args, fmt);
+    int bytes_written = minipal_log_vprint(flags, fmt, args);
+    va_end(args);
+    return bytes_written;
+}
+
+int minipal_log_vprint(minipal_log_flags flags, const char* fmt, va_list args)
+{
+    char stack_buffer[MINIPAL_LOG_BUF_SIZE];
+    int bytes_written = 0;
+    va_list args_copy;
+
+    va_copy(args_copy, args);
+
+    int len = vsnprintf(stack_buffer, sizeof(stack_buffer), fmt, args_copy);
+    if (len < sizeof(stack_buffer))
+    {
+        bytes_written = minipal_log_write(flags, stack_buffer);
+    }
+    else
+    {
+        char* dyn_buffer = (char*)malloc(len + 1);
+        if (dyn_buffer != NULL)
+        {
+            vsnprintf(dyn_buffer, len + 1, fmt, args);
+            bytes_written = minipal_log_write(flags, dyn_buffer);
+            free(dyn_buffer);
+        }
+    }
+
+    va_end(args_copy);
+    return bytes_written;
+}
+
+void minipal_log_flush(minipal_log_flags flags)
+{
+    ;
+}
+
+static size_t log_write_line(minipal_log_flags flags, const char* msg, size_t msg_len)
+{
+    char buffer[MINIPAL_LOG_MAX_PAYLOAD];
+    if (msg_len < MINIPAL_LOG_MAX_PAYLOAD)
+    {
+        strncpy(buffer, msg, msg_len);
+        buffer[msg_len] = '\0';
+        return log_write(flags, buffer, msg_len);
+    }
+
+    const char* msg_end = msg + msg_len;
+    size_t bytes_written = 0;
+    while (msg < msg_end)
+    {
+        ptrdiff_t chunk_size = MINIPAL_LOG_MAX_PAYLOAD - 1;
+        if (msg_end - msg < chunk_size)
+        {
+            chunk_size = msg_end - msg;
+        }
+
+        strncpy(buffer, msg, chunk_size);
+        buffer[chunk_size] = '\0';
+        bytes_written += log_write(flags, buffer, chunk_size);
+        msg += chunk_size;
+    }
+
+    return bytes_written;
+}
+
+int minipal_log_write(minipal_log_flags flags, const char* msg)
+{
+    if (msg == NULL || msg[0] == '\0')
+        return 0;
```

Comment:
Is this actually happening? If so, it sounds like we are failing to log what was intended in some way. Can we change this to an assert instead?

Reply:
jupp, we can assert this scenario.